kong/pdk/tracing.lua

```diff
@@ -0,0 +1,488 @@
+  end
+
+  if self.events == nil then
+    self.events = tablepool_fetch(POOL_EVENTS, 4, 0)
+    self.events[0] = 0
+  end
+
+  local obj = new_tab(0, 3)
+  obj.name = name
+  obj.time_ns = time_ns or ffi_time_unix_nano()
+
+  if attributes then
+    obj.attributes = attributes
+  end
+
+  local len = self.events[0] + 1
+  self.events[len] = obj
+  self.events[0] = len
+end
+
+--- Adds an error event to a Span
+--
+-- @function span:record_error
+-- @tparam string err error string
+function span_mt:record_error(err)
+  if type(err) ~= "string" then
+    err = tostring(err)
+  end
+
+  self:add_event("exception", {
+    ["exception.message"] = err,
+  })
+end
+
+--- Adds an error event to a Span
+-- Status codes:
+-- - `0` unset
+-- - `1` ok
+-- - `2` error
+--
+-- @function span:set_status
+-- @tparam number status status code
+function span_mt:set_status(status)
+  if type(status) ~= "number" then
+    error("invalid status", 2)
+  end
+
+  self.status = status
+end
+
+-- (internal) Release a span
+-- The lifecycle of span is controlled by Kong
+function span_mt:release()
+  if type(self.attributes) == "table" then
+    tablepool_release(POOL_ATTRIBUTES, self.attributes)
+  end
+
+  if type(self.events) == "table" then
+    tablepool_release(POOL_EVENTS, self.events)
+  end
+
+  -- metabale will be cleared
+  tablepool_release(POOL_SPAN, self)
+end
+
+-- (internal) compatible with Zipkin tracing headers
+-- TODO: implement baggage API
+function span_mt:each_baggage_item() return NOOP end
+
+local tracer_mt = {}
+tracer_mt.__index = tracer_mt
+
+-- avoid creating multiple tracer with same name
+local tracer_memo = setmetatable({}, { __mode = "k" })
+
+local noop_tracer = {}
+noop_tracer.name = "noop"
+noop_tracer.start_span = function() return noop_span end
+noop_tracer.active_span = NOOP
+noop_tracer.set_active_span = NOOP
+noop_tracer.process_span = NOOP
+
+--- New Tracer
+local function new_tracer(name, options)
+  name = name or "default"
+
+  if tracer_memo[name] then
+    return tracer_memo[name]
+  end
+
+  local self = {
+    -- instrumentation library name
+    name = name,
+  }
+
+  options = options or {}
+  if options.noop then
+    return noop_tracer
+  end
+
+  options.sampling_rate = options.sampling_rate or 1.0
+  self.sampler = get_trace_id_based_sampler(options.sampling_rate)
+  self.active_span_key = name .. "_" .. "active_span"
+
+  --- Get the active span
+  -- Returns the root span by default
+  --
+  -- @function kong.tracing.new_span
+  -- @phases rewrite, access, header_filter, response, body_filter, log, admin_api
+  -- @treturn table span
+  function self.active_span()
+    if not base.get_request() then
+      return
+    end
+
+    return ngx.ctx[self.active_span_key]
+  end
+
+  --- Set the active span
+  --
+  -- @function kong.tracing.new_span
+  -- @phases rewrite, access, header_filter, response, body_filter, log, admin_api
+  -- @tparam table span
+  function self.set_active_span(span)
```

Comment:
Would you add some comments to explain why we don't do this in `start_span` automatically?

Reply:
1. We have to do the checking for the `ctx.active_span` to see if it exists, it could be expensive.
2. At most of the time, the caller will not set the active span.